### PR TITLE
feat(ingest/tableau): Fetch Upstreams From Columns

### PR DESCRIPTION
**Pull request by egemenberk**
This PR fixes the following two things:
- Tableau metadata graphql API returns empty list for `upstreamTables` for embedded datasources while `upstreamColumns` field includes information. This PR populates upstream table information from `upstreamColumns` field coming from the Tableau
- Tableau metadata graphql API returns malfunctioned SQL queries which causing failures in fetching upstream Lineage from the CustomSQLs. This also enables embedded data sources to be connected to the CustomSQLs which are (generally) connected to upstreams from other platforms hence completing the full lineage.

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Review by asikowitz — CHANGES_REQUESTED** [review]
Thanks for this contribution! I have a couple style requests. Can you also update `tests/integration/tableau/setup/embeddedDatasourcesConnection_all.json` to add a data source that has no upstream tables, but still has lineage information in upstream columns, to test this new functionality?

**Comment by hsheth2** [comment]
@egemenberk some query cleaning logic is also getting adding in this PR https://github.com/datahub-project/datahub/pull/9838 - that one also removes parameter names and things to make SQL parsing work. Does it make sense to unify across these two query cleaning implementations?

**Comment by egemenberk** [comment]
> @egemenberk some query cleaning logic is also getting adding in this PR #9838 - that one also removes parameter names and things to make SQL parsing work. Does it make sense to unify across these two query cleaning implementations?

Hi @hsheth2, I've taken a quick look at the PR you mentioned and it seems to fix the query, so I can remove the clean_query() method call from my implementation. My PR's main focus is to fix fetching upstream lineage from `upstreamColumns` when `upstreamTables` field is empty in Tableau response, the clean_query addition was a side fix while working on the task, so I can remove my addition on that and trust the #9838 implementation. Thanks for the information 👍

File: metadata-ingestion/src/datahub/ingestion/source/tableau.py [hunk 1]
```diff
@@ -1954,6 +1954,42 @@ def emit_datasource(
                     aspect_name=c.UPSTREAM_LINEAGE,
                     aspect=upstream_lineage,
                 )
+        else:
+            # Fetch upstream information from upstreamColumns which are mostly coming
+            # from CUSTOM SQLs when Tableu doesn't provide upstream tables
+            # With this approach, we also join lingering CustomSQL entities to Embedded Data Sources
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/tableau.py:1960]: copying in your PR description

```suggestion
            # Tableau's metadata graphql API sometimes returns empty an list for upstreamTables
            # for embedded datasources. However, the upstreamColumns field often includes information.
            # This attempts to populate upstream table information from the upstreamColumns field.
```

File: metadata-ingestion/src/datahub/ingestion/source/tableau.py [hunk 2]
```diff
@@ -1954,6 +1954,42 @@ def emit_datasource(
...
+                and column.get(c.TABLE, {}).get(c.ID)
+            }
+            fine_grained_lineages = self.get_upstream_columns_of_fields_in_datasource(
+                datasource, datasource_urn, table_id_to_urn
+            )
+            upstream_tables = [
+                Upstream(dataset=table_urn, type=DatasetLineageType.TRANSFORMED)
+                for table_urn in table_id_to_urn.values()
+            ]
+            upstream_lineage = UpstreamLineage(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/tableau.py:1980]: let's put this new logic in the `_create_upstream_table_lineage` method

we can move this if statement into the method as well if necessary
```
        # Upstream Tables
        if (
            datasource.get(c.UPSTREAM_TABLES)
            or datasource.get(c.UPSTREAM_DATA_SOURCES)
            or datasource.get(c.FIELDS)
        ):
```

Comment (egemenberk) [reply metadata-ingestion/src/datahub/ingestion/source/tableau.py]: Makes sense

**Pull request merged**